iio/iio_ad7124/iio_ad7124.c

```diff
@@ -0,0 +1,1061 @@
+		size_t bytes_count,
+		uint32_t ch_mask)
+{
+	struct iio_ad7124_desc	*iio_desc = (struct iio_ad7124_desc *)iio_inst;
+	struct ad7124_dev	*desc = (struct ad7124_dev *)iio_desc->ad7124_driver_handler;
+	ssize_t ret, sample_no;
+	uint8_t ch_bulk = 3 * hweight8(ch_mask);
+	int32_t data_temp;
+	uint32_t i, j = 0, chan_idx;
+
+	if ((bytes_count % ch_bulk) != 0)
+		return -EINVAL;
+
+	sample_no = bytes_count / ch_bulk;
+
+	while (1) {
+		for (i = 0; i < hweight8(ch_mask); i++) {
+			ret = ad7124_wait_for_conv_ready(desc, 100000);
+			if (ret != SUCCESS)
+				return -ETIMEDOUT;
+			ret = ad7124_read_data(desc, &data_temp);
+			if (ret != SUCCESS)
+				return -EIO;
+			chan_idx = desc->regs[AD7124_Status].value &
+				   AD7124_STATUS_REG_CH_ACTIVE(0xF);
+			if ((chan_idx != 0) && (i == 0)) {
+				i--;
+				continue;
+			}
+			iio_desc->ddr_base_addr[j + i * 3] = *((uint8_t *)&data_temp + 1);
+			iio_desc->ddr_base_addr[j + i * 3 + 1] = *((uint8_t *)&data_temp + 2);
+			iio_desc->ddr_base_addr[j + i * 3 + 2] = *((uint8_t *)&data_temp + 3);
+		}
+		j++;
+		if (j == sample_no)
+			break;
+	}
+
+	return bytes_count;
+}
+
+/**
+ * @brief Read chunk of data from RAM to pbuf.
+ * Call "iio_axi_adc_transfer_dev_to_mem" first.
+ * This function is probably called multiple times by libtinyiiod after a
+ * "iio_axi_adc_transfer_dev_to_mem" call, since we can only read "bytes_count"
+ * bytes at a time.
+ * @param iio_inst - Physical instance of a iio_axi_adc device.
+ * @param pbuf - Buffer where value is stored.
+ * @param offset - Offset to the remaining data after reading n chunks.
+ * @param bytes_count - Number of bytes to read.
+ * @param ch_mask - Opened channels mask.
+ * @return bytes_count or negative value in case of error.
+ */
+static ssize_t iio_ad7124_read_dev(void *iio_inst, char *pbuf, size_t offset,
+				   size_t bytes_count, uint32_t ch_mask)
+{
+	struct iio_ad7124_desc	*iio_desc = (struct iio_ad7124_desc *)iio_inst;
+	uint32_t i, j = 0, current_ch = 0;
+	uint32_t *pbuf32;
+	size_t samples;
+	uint8_t ch_bulk = 3 * hweight8(ch_mask);
+
+	if (!iio_inst)
+		return FAILURE;
+
+	if (!pbuf)
+		return FAILURE;
+
+	if ((bytes_count % ch_bulk) != 0)
+		return -EINVAL;
+
+	pbuf32 = (uint32_t*)pbuf;
+	samples = bytes_count / ch_bulk;
+
+	offset = (offset * iio_desc->iio_ad7124_device->num_ch) / hweight8(ch_mask);
+
+	for (i = 0; i < samples * iio_desc->iio_ad7124_device->num_ch; i++) {
+		if (ch_mask & BIT(current_ch)) {
+			pbuf32[j] = *(uint16_t*)(iio_desc->ddr_base_addr + offset + i * 2);
+			j++;
+		}
+
+		if (current_ch + 1 < iio_desc->iio_ad7124_device->num_ch)
+			current_ch++;
+		else
+			current_ch = 0;
+	}
+
+	return bytes_count;
+}
+
+/**
+ * @brief Initialize the AD7124 IIO driver.
+ * @param [out] desc - Pointer to the driver handler.
+ * @param [in] init_param - Pointer to the initialization structure.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t iio_ad7124_init(struct iio_ad7124_desc **desc,
```

Comment:
This function is unnecesary.
Create a structure like: https://github.com/analogdevicesinc/no-OS/blob/5a93b5a6e9fb132fd92a1bef0050eb27d5e32aa8/iio/iio_demo/iio_demo_dev.h#L145